### PR TITLE
#47 add support for gdeh0154z90

### PR DESCRIPTION
**Pull request by thebino**

### CODE REVIEW

**Comment by martinberlin** [comment]
Looking good. Tell me on the next days when you finish testing and I merge it.
Thanks for your contribution!

**Review by martinberlin — CHANGES_REQUESTED** [review]
I will review this in some minutes can you please just make a photo, if you can on the window, with some light so I can add it to the Wiki. 
**updateToWindow** method can go away completely both from .h and cpp

I'm thinking that maybe it's not a bad idea to remove all partial update completely from models that do not support it. Thanks a lot!

File: components/CalEPD/include/gdeh0154z90.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    uint8_t colors_supported = 2;
+    uint8_t partial_update_supported = 0;
+
+    void init(bool debug);
+    void fillScreen(uint16_t color);
+    void drawPixel(int16_t x, int16_t y, uint16_t color); // Override GFX own drawPixel method
+
+    void update();
+    void updateWindow(uint16_t x, uint16_t y, uint16_t w, uint16_t h, bool using_rotation = true);
+    void updateToWindow(uint16_t xs, uint16_t ys, uint16_t xd, uint16_t yd, uint16_t w, uint16_t h, bool using_rotation = true);
```
Comment (martinberlin) [inline components/CalEPD/include/gdeh0154z90.h:40]: updateToWindow -> can be removed completely

**Review by martinberlin — CHANGES_REQUESTED** [review]
Overall looking great. Only this two small things: 
1. The empty IF condition, that could be simply eliminated.
2. The updateToWindow method, that I think was a leftover from gxEPD methods, and it's only in some old classes. 

But looks very good and I'm very happy that you got it working and at the same time got to know Cale-idf better.

File: components/CalEPD/include/gdeh0154z90.h (hunk 1 above)
Comment (thebino) [reply components/CalEPD/include/gdeh0154z90.h:40]: I added it in preparation of collating a general interface used by the Epd class.

File: components/CalEPD/models/gdeh0154z90.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        y = GDEH0154Z90_HEIGHT - y - 1;
+        break;
+    }
+    uint16_t i = x / 8 + y * GDEH0154Z90_WIDTH / 8;
+
+    _black_buffer[i] = (_black_buffer[i] & (GDEH0154Z90_8PIX_WHITE ^ (1 << (7 - x % 8)))); // white
+    _red_buffer[i] = (_red_buffer[i] & (GDEH0154Z90_8PIX_RED_WHITE ^ (1 << (7 - x % 8)))); // white
+
+    // In this display controller RAM colors are inverted: WHITE RAM(BW) = 1  / BLACK = 0
+    switch (color)
```
Comment (martinberlin) [inline components/CalEPD/models/gdeh0154z90.cpp:86]: If you already set both buffers stop there and do anything more. I don't think it needs further transformations

File: components/CalEPD/models/gdeh0154z90.cpp [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+void Gdeh0154z90::drawPixel(int16_t x, int16_t y, uint16_t color)
+{
+    if ((x < 0) || (x >= width()) || (y < 0) || (y >= height()))
+    {
+        return;
+    }
+
+    // check rotation, move pixel around if necessary
+    switch (getRotation())
```
Comment (martinberlin) [inline components/CalEPD/models/gdeh0154z90.cpp:65]: You can replace this whole switch using OOP with just one line:
_rotate(x, y, with(), height());

**Review by martinberlin — APPROVED** [review]

**Comment by martinberlin** [comment]
Thanks for your contribution

**Pull request merged**